File: deploy/aws/src/glow_deploy/gui/routes/heartbeat.py
"""Liveness ping from the open browser tab/window.

`main.py`'s watcher thread quits the server once heartbeats stop arriving —
the closest a browser-tab app gets to "the window was closed". No auth: it
carries no data, runs on every page (including signin), and only touches
process lifecycle.
"""

from __future__ import annotations

import time

from fastapi import APIRouter, Request
from fastapi.responses import Response

router = APIRouter()


@router.post("/heartbeat")
def heartbeat(request: Request) -> Response:
    request.app.state.last_heartbeat = time.time()
    return Response(status_code=204)

File: deploy/aws/src/glow_deploy/gui/frontend/heartbeat.ts
// Pings the server every few seconds so it can tell "tab/window closed" from
// "user is just reading this page" — see routes/heartbeat.py and main.py's
// watcher thread, which quits the process once pings stop arriving.
const PING_INTERVAL_MS = 3000;

function ping(): void {
  void fetch("/heartbeat", { method: "POST", keepalive: true }).catch(() => {});
}

export function init(): void {
  ping();
  window.setInterval(ping, PING_INTERVAL_MS);
}

if (!(globalThis as { __TEST__?: boolean }).__TEST__) init();
